contrib/gorm.io/gorm/gorm.go

```diff
@@ -0,0 +1,181 @@
+
+	sqltraced "gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+
+	"gorm.io/gorm"
+)
+
+const (
+	gormConfigKey        = "dd-trace-go:config"
+	gormSpanStartTimeKey = "dd-trace-go:span"
+)
+
+// Open opens a new (traced) database connection. The used driver must be formerly registered
+// using (gopkg.in/DataDog/dd-trace-go.v1/contrib/database/sql).Register.
+func Open(getDialector func(db *sql.DB) gorm.Dialector, driverName, source string, opts ...Option) (*gorm.DB, error) {
+	dialector := getDialector(nil)
+
+	sqldb, err := sqltraced.Open(driverName, source)
+	if err != nil {
+		return nil, err
+	}
+
+	dialector = getDialector(sqldb)
+
+	db, err := gorm.Open(dialector, &gorm.Config{})
+	if err != nil {
+		return db, err
+	}
+
+	return WithCallbacks(db, opts...)
+}
+
+// WithCallbacks registers callbacks to the gorm.DB for tracing.
+// It should be called once, after opening the db.
+// The callbacks are triggered by Create, Update, Delete,
+// Query and RowQuery operations.
+func WithCallbacks(db *gorm.DB, opts ...Option) (*gorm.DB, error) {
+	afterFunc := func(operationName string) func(*gorm.DB) {
+		return func(db *gorm.DB) {
+			after(db, operationName)
+		}
+	}
+
+	cfg := new(config)
+	defaults(cfg)
+	for _, fn := range opts {
+		fn(cfg)
+	}
+
+	ctx := context.Background()
+	if db.Statement != nil && db.Statement.Context != nil {
+		ctx = db.Statement.Context
+	}
+
+	db = db.WithContext(context.WithValue(ctx, gormConfigKey, cfg))
+
+	cb := db.Callback()
+	err := cb.Create().Before("gorm:create").Register("dd-trace-go:before_create", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Create().After("gorm:create").Register("dd-trace-go:after_create", afterFunc("gorm.create"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Update().Before("gorm:update").Register("dd-trace-go:before_update", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Update().After("gorm:update").Register("dd-trace-go:after_update", afterFunc("gorm.update"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Delete().Before("gorm:delete").Register("dd-trace-go:before_delete", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Delete().After("gorm:delete").Register("dd-trace-go:after_delete", afterFunc("gorm.delete"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Query().Before("gorm:query").Register("dd-trace-go:before_query", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Query().After("gorm:query").Register("dd-trace-go:after_query", afterFunc("gorm.query"))
+	if err != nil {
+		return db, err
+	}
+	err = cb.Row().Before("gorm:query").Register("dd-trace-go:before_row_query", before)
+	if err != nil {
+		return db, err
+	}
+	err = cb.Row().After("gorm:query").Register("dd-trace-go:after_row_query", afterFunc("gorm.row_query"))
+	if err != nil {
+		return db, err
+	}
+
+
+
```

Comment:
Please remove most empty lines. The style convention in this repository is mostly following the Go standard library. You will notice that most of the code is compact and uses very little empty lines (if at all)